lib/mongo/auth/stringprep.rb

```diff
@@ -0,0 +1,68 @@
+require 'mongo/auth/stringprep/tables'
+
+module Mongo
+  module Auth
+    module StringPrep
+      extend self
+
+      def prepare(data, mappings, prohibited, options = {})
+        out = apply_maps(data, mappings)
+        normalize(out) if options[:normalize]
+        check_prohibited(out, prohibited)
+        check_bidi(out) if options[:bidi]
+        out
+      end
+
+      private
+
+      def apply_maps(data, mappings)
+        out = ''
+
+        data.each_char do |c|
+          out << mapping(c, mappings)
+        end
+
+        out
+      end
+
+      def check_bidi(out)
+        if out.any? { |c| table_contains?(Tables::C8, c) }
+          raise Error::StringPrep(Error::StringPrep::INVALID_BIDIRECTIONAL)
+        end
+
+        if out.any? { |c| table_contains?(Tables::D1, c) }
+          unless table_contains?(Tables::D1, out.first) && table_contains?(Tables::D1, out.last)
+            raise Error::StringPrep(Error::StringPrep::INVALID_BIDIRECTIONAL)
+          end
+        end
+      end
+
+      def check_prohibited(out, prohibited)
+        out.any? do |c|
+          prohibited.any? do |table|
```

Comment:
Same question about any

Reply:
The algorithm is that if any character in the string is in any of the tables in the list of prohitibted tables, then the string is invalid. `any?` seemed like the most straightforward way to implement this.